weave/index/index_sql.py

```diff
@@ -0,0 +1,869 @@
+
+        self._file_system = file_system
+        self._pantry_path = pantry_path
+
+        self._database_name = kwargs.get("database_name", "weave_db")\
+            .replace("-", "_")
+        d_schema_name = self._pantry_path.replace(os.sep, "_")\
+            .replace("-", "_")
+        self._pantry_schema = kwargs.get("pantry_schema", d_schema_name)
+
+        # Make a connection to tempdb, then create the desired database if it
+        # does not exist. Then connect to the desired database, and create the
+        # tables if they do not exist.
+        self._connect(database_name="tempdb", **kwargs)
+        self._select_db(database_name=self._database_name)
+        self._create_schema()
+        self._create_tables()
+
+    # Create read-only properties for the database and schema names.
+    @property
+    def database_name(self):
+        """The database name of the index."""
+        return self._database_name
+
+    @property
+    def pantry_schema(self):
+        """The schema name of the index."""
+        return self._pantry_schema
+
+    def __del__(self):
+        """Close the connection to the database."""
+        self.cur.close()
+        self.con.close()
+
+    def _connect(self, database_name="weave_db", **kwargs):
+        """Connect to the server, and select the specified DB.
+
+        Parameters
+        ----------
+        database_name: str (default='weave_db')
+            The database to connect to.
+        **driver: str (default='{ODBC Driver 18 for SQL Server}')
+            Driver to be used for the connection. Usually similar to:
+            '{ODBC Driver <18, 17, 13...> for SQL Server}'.
+        **encrypt: bool (default=False)
+            Whether or not to encrypt the database (disabled by default, due
+            to the fact that it requires additional configuration).
+        """
+        driver = kwargs.get("odbc_driver", "{ODBC Driver 18 for SQL Server}")
+        encrypt = kwargs.get("encrypt", False)
+
+        # Pylint has a problem recognizing 'connect' as a valid member function
+        # so we ignore that here.
+        # pylint: disable-next=c-extension-no-member
+        self.con = pyodbc.connect(
+            f"DRIVER={driver};"
+            f"SERVER={os.environ['MSSQL_HOST']};"
+            f"DATABASE={database_name};"
+            f"UID={os.environ['MSSQL_USERNAME']};"
+            f"PWD={os.environ['MSSQL_PASSWORD']};"
+            f"Encrypt={'yes' if encrypt else 'no'};"
+        )
+        self.cur = self.con.cursor()
+
+    def _select_db(self, database_name="weave_db"):
+        """Select the database to use, or create it if it does not exist.
+
+        Parameters
+        ----------
+        database_name: str (default='weave_db')
+            The database to connect to.
+        """
+        # Query to see if the database exists.
+        self.cur.execute(
+            "SELECT * FROM sys.databases WHERE name = ?", (database_name,)
```

Comment:
is the "?" a placeholder for variables that you input?

Reply:
Yes, this is how MS SQL handles parameterized values